llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,408 @@
+    if (isIdentityCopy(MI))
+      continue;
+
+    for (const MachineOperand &MO : MI.defs()) {
+
+      Register Reg = MO.getReg();
+      if (!Reg.isVirtual())
+        continue;
+      if (VRM->hasRequiredPhys(Reg))
+        continue;
+      if (MO.isTied())
+        continue;
+      // several definitions of the same virtual register are not relevant
+      // because even if the virtual register is renamed, by construction
+      // all the definitions would be renamed as well and achieve nothing wrt
+      // WAW dependecy resolution
+      if (LastVRegDef[Reg] != &MI)
+        continue;
+
+      if (isWorthRenaming(Reg, UsedPhysRegs, VRegWithCopies) &&
+          replaceReg(Reg, BlockedPhysRegs)) {
+
+        LLVM_DEBUG(dbgs() << MI);
+        Modified = true;
+
+      } else {
+        // Keep track of already visited physical registers.
+        // Incrementally add the encountered physical registers, so that a
+        // second occurrence of a physical register can trigger the register
+        // rewriting
+        // Blocked registers for the replacement are recorded in
+        // BlockedPhysRegs. Initially all the used physical registers
+        // from the MBB are blocked, so that replacements do not introduce WAW
+        // dependencies. Additionally, replaced registers are already blocked in
+        // BlockedPhysRegs, so that an additional replacement will not cause a
+        // WAW, which this pass is trying to remove.
+        UsedPhysRegs[VRM->getPhys(Reg)] = true;
+      }
+    }
+  }
+
+  return Modified;
+}
+
+bool AIEWawRegRewriter::isWorthRenaming(const Register &Reg,
+                                        const BitVector &UsedPhysRegs,
+                                        const BitVector &VRegWithCopies) const {
+  assert(Reg.isVirtual());
+
+  // Only rename registers mapped to a phys reg assigned more than once
+  if (!UsedPhysRegs[VRM->getPhys(Reg)])
+    return false;
+
+  if (!TRI->isVecOrAccRegClass(*(MRI->getRegClass(Reg))))
+    return false;
+
+  return !VRegWithCopies[Reg.virtRegIndex()];
+}
+
+BitVector
+AIEWawRegRewriter::getDefinedPhysRegs(const MachineBasicBlock *MBB) const {
+  BitVector BlockedPhysRegs(TRI->getNumRegs());
+
+  for (const MachineInstr &MI : *MBB) {
+    for (const MachineOperand &Op : MI.defs()) {
+      MCPhysReg PhysReg = getAssignedPhysReg(Op.getReg());
+      if (MCRegister::isPhysicalRegister(PhysReg))
+        addAliasRegs(BlockedPhysRegs, PhysReg);
+    }
+  }
+
+  return BlockedPhysRegs;
+}
+
+bool AIEWawRegRewriter::replaceReg(const Register Reg,
+                                   BitVector &BlockedPhysRegs) {
+  assert(Reg.isVirtual());
+  LLVM_DEBUG(dbgs() << " WAW RegRewriter: Register to replace "
+                    << TRI->getName(VRM->getPhys(Reg)) << "\n");
+
+  MCPhysReg ReplacementPhysReg = getReplacementPhysReg(Reg, BlockedPhysRegs);
+
+  if (!MCRegister::isPhysicalRegister(ReplacementPhysReg))
+    return false;
```

Comment:
Super-nit: I think it means we could not find a replacement register. So we could make the check more explicit by either:
 - Having `getReplacementPhysReg` return a `std::optional`
 - Or checking `ReplacementPhysReg == MCRegister::NoRegister` explicitely